File: 50dias/python/28.py
import unicodedata

def RemovePontuacao(texto):
    textoEditado = texto.replace(' ', '')
    textoEditado = textoEditado.replace('.', '')
    textoEditado = textoEditado.replace(',', '')
    textoEditado = textoEditado.replace('!', '')
    textoEditado = textoEditado.replace('?', '')
    textoEditado = textoEditado.replace('-', '')
    textoEditado = textoEditado.replace(':', '')
    textoEditado = textoEditado.replace(';', '')
    return textoEditado

def RemoveAcentuacao(texto):
    normalized = unicodedata.normalize('NFD', texto)
    return normalized.encode('ascii', 'ignore').decode('utf8').casefold()

def NormalizaTexto(texto):
    textoEditado = RemovePontuacao(texto)
    textoEditado = RemoveAcentuacao(textoEditado)
    return textoEditado

def EhPalindromo(texto):
    texto = NormalizaTexto(texto)
    inicioTexto = 0
    fimTexto = len(texto) - 1

    while inicioTexto < fimTexto:
        if texto[inicioTexto] != texto[fimTexto]:
            return False
        inicioTexto += 1
        fimTexto -= 1
    return True

def main():
    texto = input('Digite um texto: ')
    ehPalindromo = EhPalindromo(texto)

    naoEh = "" if ehPalindromo else "não "
    mensagem = "O texto {}é palíndromo!".format(naoEh)
    print(mensagem)

main()

File: 50dias/cpp/28.cpp
#include <iostream>
#include <algorithm>
#include <string>

using namespace std;

void LeTexto(string& texto) {
	cout << "Digite um texto: ";
	getline(cin, texto);
}

void NormalizaTexto(string& texto) {
	erase(texto, ' ');
	erase(texto, '.');
	erase(texto, ',');
	erase(texto, '!');
	erase(texto, '?');
	erase(texto, ':');
	erase(texto, ';');
	erase(texto, '-');
	std::for_each(texto.begin(), texto.end(), [](char& c) {
		c = ::tolower(c);
		});
}

bool EhPalindromo(string texto) {
	int inicioTexto = 0;
	int fimTexto = texto.size() - 1;

	while (inicioTexto < fimTexto) {
		if (texto[inicioTexto] != texto[fimTexto])
			return false;
		inicioTexto++;
		fimTexto--;
	}

	return true;
}

int main() {
	string texto, textoEditado;

	LeTexto(texto);
	textoEditado = texto;
	NormalizaTexto(textoEditado);
	bool palindromo = EhPalindromo(textoEditado);

	cout << "-------------------------" << endl;
	cout << "    " << texto << endl;
	cout << "-------------------------" << endl;
	cout << "Texto ";
	if (!palindromo)
		cout << "nao ";
	cout << "e palindromo!" << endl << endl;

	return 0;
}